src/asm/warning.c

```diff
@@ -0,0 +1,244 @@
+	WARNING_USER,
+	WARNING_OBSOLETE,
+	WARNING_BUILTIN_ARG,
+	WARNING_LARGE_CONSTANT,
+	WARNING_SHIFT,
+	WARNING_DIV,
+	WARNING_EMPTY_ENTRY,
+	WARNING_LONG_STR,
+	META_WARNING_DONE
+};
+
+static uint8_t const *metaWarningCommands[NB_META_WARNINGS] = {
+	_wallCommands,
+	_wextraCommands,
+	_weverythingCommands
+};
+
+void processWarningFlag(char const *flag)
+{
+	static bool setError = false;
+
+	/* First, try to match against a "meta" warning */
+	for (enum WarningID id = NB_WARNINGS; id < NB_WARNINGS_ALL; id++) {
+		/* TODO: improve the matching performance? */
+		if (!strcmp(flag, warningFlags[id])) {
+			/* We got a match! */
+			uint8_t const *ptr =
+					metaWarningCommands[id - NB_WARNINGS];
+
+			for (;;) {
+				switch (*ptr) {
+				case META_WARNING_DONE:
+					return;
+
+				default:
+					/* Warning flag, set without override */
+					if (warningStates[*ptr]
+					 == WARNING_DEFAULT)
+						warningStates[*ptr] =
+								WARNING_ENABLED;
+				}
+				ptr++;
+			}
+		}
+	}
+
+	/* If it's not a meta warning, specially check against `-Werror` */
+	if (!strncmp(flag, "error", strlen("error"))) {
+		char const *errorFlag = flag + strlen("error");
+
+		switch (*errorFlag) {
+		case '\0':
+			/* `-Werror` */
+			warningsAreErrors = true;
+			return;
+
+		case '=':
+			/* `-Werror=XXX */
+			setError = true;
+			processWarningFlag(errorFlag + 1); /* Skip the `=` */
+			setError = false;
+			return;
+
+		/* Otherwise, allow parsing as another flag */
+		}
+	}
+
+	/* Well, it's either a normal warning or a mistake */
+
+	/* Check if this is a negation */
+	bool isNegation = !strncmp(flag, "no-", strlen("no-")) && !setError;
+	char const *rootFlag = isNegation ? flag + strlen("no-") : flag;
+	enum WarningState state = setError ? WARNING_ERROR :
+			isNegation ? WARNING_DISABLED : WARNING_ENABLED;
+
+	/* Try to match the flag against a "normal" flag */
+	for (enum WarningID id = 0; id < NB_WARNINGS; id++) {
+		if (!strcmp(rootFlag, warningFlags[id])) {
+			/* We got a match! */
+			warningStates[id] = state;
+			return;
+		}
+	}
+
+	warnx("Unknown warning `%s`", flag);
+}
+
+void verror(const char *fmt, va_list args, char const *flag)
+{
+	fputs("ERROR: ", stderr);
+	fstk_Dump();
+	fprintf(stderr, flag ? ": [-Werror=%s]\n    " : ":\n    ", flag);
+	vfprintf(stderr, fmt, args);
+	fputc('\n', stderr);
+	fstk_DumpStringExpansions();
+	nbErrors++;
+}
+
+void yyerror(const char *fmt, ...)
+{
+	va_list args;
+
+	va_start(args, fmt);
+	verror(fmt, args, NULL);
+	va_end(args);
+}
+
+noreturn_ void fatalerror(const char *fmt, ...)
+{
+	va_list args;
+
+	va_start(args, fmt);
+	verror(fmt, args, NULL);
+	va_end(args);
+
+	exit(5);
+}
+
+void warning(enum WarningID id, char const *fmt, ...)
+{
+	char const *flag = warningFlags[id];
+	va_list args;
+
+	va_start(args, fmt);
+
+	switch (warningState(id)) {
+	case WARNING_DISABLED:
+		return;
+
+	case WARNING_ERROR:
+		verror(fmt, args, flag);
+		va_end(args);
+		return;
+
+	case WARNING_DEFAULT:
```

Comment:
Don't add a `case` label for an unreachable `case` — remember that execution simply continues at the next statement (labels are ignored; in other words, execution falls through). If you must handle this value, use something like GCC's `__builtin_unreachable()` (or even `abort()`) — but it's best to not handle it at all; add a comment if needed. Adding a `case` label with incorrect behavior is worse than nothing.

Reply:
Removing the label causes a warning due to not handling an enum case. Adding an "unreachable" would be better, yes.

Reply:
Added a call to `abort()` locally, waiting for other conversations to be resolved before pushing